### PR TITLE
[ERC721] Fix pre-conditions based on EIP

### LINKED ISSUE #24: [ERC721] Unnecessary zero address check at `_validateTransferFrom`
https://github.com/LayerXcom/verified-vyper-contracts/blob/fca925befdc4c5c296c95e8df99db31426aeac48/erc721/ERC721.vy#L177

### PR DESCRIPTION
**Pull request by nrryuya**
resolve #24 (and others)

### CODE REVIEW

**Comment by nrryuya** [comment]
Waiting for #18 #23

File: erc721/ERC721.vy [hunk 1]
```diff
@@ -249,64 +245,76 @@ def safeTransferFrom(
...
+#      Throws if `_tokenId` is not a valid NFT.
 # @param tokenId uint256 id of the ERC721 token to be burned.
 @public
 def burn(_tokenId: uint256):
+    # Check requirements
     assert self._isApprovedOrOwner(msg.sender, _tokenId)
-    owner: address = self._ownerOf(_tokenId)
+    owner: address = self.idToOwner[_tokenId]
+    # Throws if `_tokenId` is not a valid NFT
+    assert owner == ZERO_ADDRESS
```
Comment (yudetamago) [inline erc721/ERC721.vy]: Is this intended to check `assert owner != ZERO_ADDRESS` ?

Comment (nrryuya) [reply erc721/ERC721.vy]: :dogeza:

File: erc721/ERC721.vy [hunk 2]
```diff
@@ -249,64 +245,76 @@ def safeTransferFrom(
...
 
 
 # @dev Enables or disables approval for a third party ("operator") to manage all of
 #      `msg.sender`'s assets. It also emits the ApprovalForAll event.
-# @notice This works even if sender doesn't own any tokens at the time.
+#      NOTE: This works even if sender doesn't own any tokens at the time.
 # @param _operator Address to add to the set of authorized operators.
 # @param _isApproved True if the operators is approved, false to revoke approval.
 @public
 def setApprovalForAll(_operator: address, _approved: bool):
```
Comment (yudetamago) [inline erc721/ERC721.vy]: [IMO]
It is better to check `_operator != owner` at the top of `setApprovalForAll`

Comment (nrryuya) [reply erc721/ERC721.vy]: hmm... I think it's OK to approve for themselves?

Comment (yudetamago) [reply erc721/ERC721.vy]: Yes it is basically ok, but when executed (by mistake), more gas is consumed.
To approve themselves is needed?

Comment (nrryuya) [reply erc721/ERC721.vy]: OK, to prevent `approve` and `setApprovalForAll` to zero address and sender are not required in the original EIP but improve security or avoid cases the sender probably don't intend. Whether or not to add these assertions is different among implementations and it has a small trade-off of gas cost.

Comment (yudetamago) [reply erc721/ERC721.vy]: agree

Comment (nrryuya) [reply erc721/ERC721.vy]: fixed

File: erc20/spec-tmpl.k [hunk 3]
```diff
@@ -69,7 +69,7 @@
 
         <accounts>
           <account>
-            <acctID> ACCT_ID </acctID>
+            <acctID> ACCT_ID:Int </acctID>
```
Comment (yudetamago) [inline erc20/spec-tmpl.k]: This modification is intended to be included in another PR? 👀
